nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/storage/sql/CloudFrontBinaryManager.java

```diff
@@ -0,0 +1,136 @@
+
+    public static final String PRIVATE_KEY_PROPERTY = BASE_PROP + "privKey";
+
+    public static final String PRIVATE_KEY_ID_PROPERTY = BASE_PROP + "privKeyId";
+
+    public static final String DISTRIB_DOMAIN_PROPERTY = BASE_PROP + "distribDomain";
+
+    public static final String PROTOCOL_PROPERTY = BASE_PROP + "protocol";
+
+    public static final String ENABLE_CF_ENCODING_FIX = "org.nuxeo.s3storage.cloudfront.fix.encoding";
+
+    protected String distributionDomain;
+
+    protected Protocol protocol;
+
+    protected PrivateKey privKey;
+
+    protected String privKeyId;
+
+    @Override
+    protected void setupCloudClient() throws IOException {
+        super.setupCloudClient();
+
+        protocol = Protocol.valueOf(getProperty(PROTOCOL_PROPERTY, "https"));
+        distributionDomain = getProperty(DISTRIB_DOMAIN_PROPERTY);
+
+        try {
+            String privateKeyPath = getProperty(PRIVATE_KEY_PROPERTY);
+            privKey = loadPrivateKey(privateKeyPath);
+            privKeyId = getProperty(PRIVATE_KEY_ID_PROPERTY);
+        } catch (InvalidKeySpecException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    protected URI getRemoteUri(String digest, ManagedBlob blob, HttpServletRequest servletRequest) throws IOException {
+        try {
+            URIBuilder uriBuilder = new URIBuilder(buildResourcePath(bucketNamePrefix + digest));
+            if (blob != null) {
+                uriBuilder.addParameter("response-content-type", getContentTypeHeader(blob));
+                uriBuilder.addParameter("response-content-disposition",
+                        getContentDispositionHeader(blob, servletRequest));
+            }
+
+            if (isBooleanPropertyTrue(ENABLE_CF_ENCODING_FIX)) {
+                String trimmedChars = " ";
+                uriBuilder.getQueryParams().stream().filter(s -> s.getValue().contains(trimmedChars)).forEach(
+                        s -> uriBuilder.setParameter(s.getName(), s.getValue().replace(trimmedChars, "")));
+            }
+
+            URI uri = uriBuilder.build();
+            if (privKey == null) {
+                return uri;
+            }
+
+            Date expiration = new Date();
+            expiration.setTime(expiration.getTime() + directDownloadExpire * 1000);
+
+            String signedURL = CloudFrontUrlSigner.getSignedURLWithCannedPolicy(uri.toString(), privKeyId, privKey,
+                    expiration);
+            return new URI(signedURL);
+        } catch (URISyntaxException e) {
+            throw new IOException(e);
+        }
+    }
+
+    private String buildResourcePath(String s3ObjectKey) {
+        return protocol != CloudFrontUrlSigner.Protocol.http && protocol != CloudFrontUrlSigner.Protocol.https
+                ? s3ObjectKey : protocol + "://" + distributionDomain + "/" + s3ObjectKey;
+    }
+
+    private static PrivateKey loadPrivateKey(String privateKeyPath) throws InvalidKeySpecException, IOException {
```

Comment:
Why private static? It seems used to set an object's field, after the object instantiation. It seems not used in a test.